### PR TITLE
Loading cross section time series pp files (lbcode 31323)

### PR DESCRIPTION
**Pull request by lbdreyer**
This includes the changes to the pp rules to correctly load in a file with an lbcode of 31323

@marqh I have removed the logic that doesn't apply specifically to the example file I was looking at. Tests will be added.and I will fix the cdl/cml files. This is just for you to take a look at as you asked.

### CODE REVIEW

File: lib/iris/fileformats/pp_rules.py [hunk 1]
```diff
@@ -1004,18 +1014,54 @@ def _all_other_rules(f):
...
+                points = f.upper_z_domain
+            aux_coords_and_dims.append(
+                (AuxCoord(
+                    points,
+                    long_name='site_number',
+                    units='1'),
+                 1))
+        else:
+            msg = 'PP cross-section file is over-specified for the site number coordinate.'
+            raise ImportError(msg)
```
Comment (lbdreyer) [inline lib/iris/fileformats/pp_rules.py]: Wrong exception type! I will fix this

File: lib/iris/fileformats/pp_rules.py [hunk 2]
```diff
@@ -1004,18 +1014,54 @@ def _all_other_rules(f):
...
+    if (len(f.lbcode) == 5 
+            and f.lbcode[-1] == 3 
+            and f.lbcode.ix == 13 
+            and hasattr(f, 'lower_z_domain') and all(f.lower_z_domain != -1.e+30)
+            or hasattr(f, 'lower_z_domain') and all(f.lower_z_domain != -1.e+30)):
+        if f.bzx == 0 and f.bdx == 0 and not hasattr(f, 'x'):
+            if hasattr(f, 'lower_z_domain'):
+                points = f.lower_z_domain
+            if hasattr(f, 'upper_z_domain'):
+                points = f.upper_z_domain
```
Comment (lbdreyer) [inline lib/iris/fileformats/pp_rules.py]: The example file has the site number information in lower_z_domain and upper_z_domain. We can check that they are consistent with each other before preceding with using eith OR choose which one takes precedence and only use the other if the first doesn't exist. I think the first option is best.

**Comment by pelson** [comment]
Looks like there should be a cross-reference between this and #1853.

**Comment by lbdreyer** [comment]
Thanks @pelson 

This PR is also based off some work done in #1535

File: lib/iris/fileformats/pp_rules.py (hunk 2 above)
Comment (lbdreyer) [inline lib/iris/fileformats/pp_rules.py]: Thinking about it, lower_ and upper_z_domain should behave similar to lower_ and upper_x_domain, i.e. they provide the bounds for the values. So for site number, they should always be consistent.

File: lib/iris/fileformats/pp_rules.py [hunk 3]
```diff
@@ -1008,14 +1018,48 @@ def _all_other_rules(f):
...
-    if (len(f.lbcode) == 5 and f.lbcode[-1] == 1 and f.lbcode.ix == 13 and
-            f.bdx != 0):
+    if (len(f.lbcode) == 5 and f.lbcode[-1] == 1 and f.lbcode.ix == 13):
         dim_coords_and_dims.append(
-            (DimCoord.from_regular(f.bzx, f.bdx, f.lbnpt,
-                                   long_name='site_number', units='1'),
+            (DimCoord(
+                f.x,
+                long_name='site_number',
+                units='1'),
```
Comment (lbdreyer) [inline lib/iris/fileformats/pp_rules.py]: I made these changes as the documentation said that for an lbcode of 1XXXX, the x- and y-coordinate vectors must be contained in the extra data.

**Comment by lbdreyer** [comment]
**In summary:**
This PR adds 3 new rules and modifies another. 

* The 3 new rules handle the 'site_number', 'time' and 'forecast_reference_time' for a specific file with an lbcode of 31323

* The modified rule handles the 'site_number' for a file with a lbcode of 113XX to conform with the documenation

File: lib/iris/tests/unit/fileformats/pp_rules/test__convert_scalar_time_coords.py [hunk 4]
```diff
@@ -226,5 +226,22 @@ def test_unrecognised(self):
...
+        lbtim = _lbtim(ib=2, ic=2)
+        t1 = nc_datetime(1970, 1, 3, hour=0, minute=0, second=0)
+        t2 = nc_datetime(1970, 1, 4, hour=0, minute=0, second=0)
+        lbft = 24 * 4
+        coords_and_dims = _convert_scalar_time_coords(
+            lbcode=lbcode, lbtim=lbtim, epoch_hours_unit=_EPOCH_TIME_UNIT,
+            t1=t1, t2=t2, lbft=lbft)
+        t2_hours = 24 * 3
+        expected_result = [(DimCoord([t2_hours-lbft],
+                                     standard_name='forecast_reference_time',
```
Comment (marqh) [inline lib/iris/tests/unit/fileformats/pp_rules/test__convert_scalar_time_coords.py]: why is this expected result forecast_reference_time and not time please?

Comment (lbdreyer) [inline lib/iris/tests/unit/fileformats/pp_rules/test__convert_scalar_time_coords.py]: This test is for "scalar time coords". Forecast reference time is a scalar, time isn't for an lbcode of 31323

File: lib/iris/fileformats/pp_rules.py (hunk 3 above)
Comment (lbdreyer) [inline lib/iris/fileformats/pp_rules.py]: I will remove this change, I mostly added it to highlight the non-conformity to the docs.

**Comment by lbdreyer** [comment]
Possible areas of concern:
* time and forecast_reference_time have different epochs
* time has no bounds
To fix:
* The time points aren't being represented correctly e.g. value of 74880. not 1978-11-01 00:00:00
DimCoord(array([ 74880.]), standard_name='forecast_reference_time', units=Unit('hours since 1970-01-01 00:00:00', calendar='360_day'))

**Comment by marqh** [comment]
I would like to see the behaviour we currently have for reading in Station IDs restored.  I think that a clarification to UMDP F03 may be helpful to be done in parallel (i am working on this)

This would result in the remaining failing test being fixed.

File: lib/iris/fileformats/pp_rules.py [hunk 5]
```diff
@@ -1008,14 +1018,48 @@ def _all_other_rules(f):
...
+    if (len(f.lbcode) == 5
+            and f.lbcode[-1] == 3
+            and f.lbcode.ix == 13
+            and hasattr(f, 'lower_z_domain')
+            and hasattr(f, 'upper_z_domain')
+            and all(f.lower_z_domain != -1.e+30)
+            and all(f.upper_z_domain != -1.e+30)):
+        if f.bzx == 0 and f.bdx == 0 and not hasattr(f, 'x'):
+            aux_coords_and_dims.append(
+                (AuxCoord((f.upper_z_domain + f.lower_z_domain) / 2.,
```
Comment (marqh) [inline lib/iris/fileformats/pp_rules.py]: please enforce that these two numbers must be the same, rather than averaging them

**Comment by marqh** [comment]
https://code.metoffice.gov.uk/trac/um/changeset?reponame=&new=16008%40doc%2Fbranches%2Fdev%2Fmarkhedley%2Fr15746_f3timeseries&old=15746%40doc%2Ftrunk

(if you can see it)
provides suggested changes to F03 to be consistent with the implementations coded here

**Comment by lbdreyer** [comment]
I have removed the site number rule. It was incorrect to take this from the upper- and lower-z-domain as these turn out to represent the level information.
It doesn't look like the site number information has been encoded.

This leaves us with a rule for the "time" coord and the "forecast reference time" coord.

**Pull request merged**